File: java/GenerateParentheses.java
// 22. Generate Parentheses
// Accepted 3ms

class Solution {
    private List<String> parens;
    
    private void dfs(int open, int closed, String paren, final int N) {
        if(open < closed || open > N)
            return;
        
        if(open == N && closed == N) {
            parens.add(paren);
            return;
        }
        
        dfs(open + 1, closed, paren + "(", N);
        dfs(open, closed + 1, paren + ")", N);
        
        return;
    }
    
    public List<String> generateParenthesis(int n) {
        parens = new ArrayList<String>();
        dfs(0, 0, "", n);
        
        return parens;
    }
}

File: python3/022. Generate Parentheses/GenerateParentheses.py
# 22. Generate Parentheses
# Accepted 36ms

class Solution:
    def dfs(self, opened, closed, row, ret, N):
        if opened == N and closed == N:
            ret.append(row)
            return
        
        if opened < N:
            self.dfs(opened + 1, closed, row + '(', ret, N)
        if closed < opened:
            self.dfs(opened, closed + 1, row + ')', ret, N)
        
        return
    
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        ret = []
        self.dfs(0, 0, '', ret, n)
        
        return ret
